databuilder/extractor/bigquery_usage_extractor.py

```diff
@@ -0,0 +1,110 @@
+from collections import namedtuple
+from datetime import date, timedelta
+import logging
+
+from googleapiclient.discovery import build
+from oauth2client.client import GoogleCredentials
+
+from databuilder.extractor.base_extractor import Extractor
+
+TableColumnUsageTuple = namedtuple('TableColumnUsageTuple', ['database', 'cluster', 'schema',
+                                                             'table', 'column', 'email'])
+
+LOGGER = logging.getLogger(__name__)
+
+
+class BigQueryTableUsageExtractor(Extractor):
+    """
+    An aggregate extractor for bigquery table usage. This class takes the data from
+    the stackdriver logging API by filtering on timestamp, bigquery_resource and looking
+    for referencedTables in the response.
+    """
+    TIMESTAMP_KEY = 'timestamp'
+    PROJECT_ID_KEY = 'projectid'
+    DEFAULT_PAGE_SIZE = 300
+    PAGE_SIZE_KEY = 'pagesize'
+
+    def init(self, conf):
+        credentials = GoogleCredentials.get_application_default()
+        self.logging_service = build('logging', 'v2', credentials=credentials)
+
+        # type: (ConfigTree) -> None
+        self.timestamp = conf.get_string(
+            BigQueryTableUsageExtractor.TIMESTAMP_KEY,
+            (date.today() - timedelta(days=1)).strftime('%Y-%m-%dT00:00:00Z'))
+        self.projectid = conf.get_string(BigQueryTableUsageExtractor.PROJECT_ID_KEY)
+        self.pagesize = conf.get_int(
+            BigQueryTableUsageExtractor.PAGE_SIZE_KEY,
+            BigQueryTableUsageExtractor.DEFAULT_PAGE_SIZE)
+        self.table_usage_counts = {}
+        self._count_usage()
+        self.iter = iter(self.table_usage_counts)
+
+    def _count_usage(self):
+        count = 0
+        for entry in self._retrieve_records():
+            count += 1
+            if count % self.pagesize == 0:
+                LOGGER.info('Aggregated {} records'.format(count))
+
+            email = entry['protoPayload']['authenticationInfo']['principalEmail']
+            jobCompletedEvent = entry['protoPayload']['serviceData']['jobCompletedEvent']
+            refTables = jobCompletedEvent['job']['jobStatistics'].get('referencedTables', None)
+
+            if not refTables:
+                # Query results can be cached and if the source tables remain untouched,
+                # bigquery will return it from a 24 hour cache result instead.
+                continue
+
+            for refTable in refTables:
+                key = TableColumnUsageTuple(database='bigquery',
+                                            cluster=refTable['projectId'],
+                                            schema=refTable['datasetId'],
+                                            table=refTable['tableId'],
+                                            column='*',
+                                            email=email)
+                new_count = self.table_usage_counts.get(key, 0) + 1
+                self.table_usage_counts[key] = new_count
```

Comment:
we might be able to combine line 66-67 using python `setdefault()` method.

Reply:
I'm afraid not. There's an atomic operation that depends on the current value. setdefault() sets the default if it doesn't exist yet, but we always need to inc+1 to the current value, which requires 2 statements.

Reply:
Apologies, if it wasn't clear. I meant something like this 
``` self.table_usage_counts[key] = self.table_usage_counts.setdefault(key, 0) + 1```
or 
```self.table_usage_counts[key] = self.table_usage_counts.get(key, 0) + 1```

Either ways, I am fine with merging this one as is too! Feel free to resolve the comment. Thanks!